src/Microsoft.AspNetCore.JsonPatch/Internal/DynamicObjectAdapter.cs

```diff
@@ -0,0 +1,221 @@
+                return false;
+            }
+
+            errorMessage = null;
+            return true;
+        }
+
+        public bool TryRemove(
+            object target,
+            string segment,
+            IContractResolver contractResolver,
+            out string errorMessage)
+        {
+            object property = null;
+            if (!TryGetDynamicObjectProperty(target, contractResolver, segment, out property))
+            {
+                errorMessage = Resources.FormatTargetLocationAtPathSegmentNotFound(segment);
+                return false;
+            }
+
+            // Setting the value to "null" will use the default value in case of value types, and
+            // null in case of reference types
+            object value = null;
+            if (property.GetType().GetTypeInfo().IsValueType
+                && Nullable.GetUnderlyingType(property.GetType()) == null)
+            {
+                value = Activator.CreateInstance(property.GetType());
+            }
+
+            if (!TrySetDynamicObjectProperty(target, contractResolver, segment, value))
+            {
+                errorMessage = Resources.FormatTargetLocationAtPathSegmentNotFound(segment);
+                return false;
+            }
+
+            errorMessage = null;
+            return true;
+
+        }
+
+        public bool TryReplace(
+            object target,
+            string segment,
+            IContractResolver contractResolver,
+            object value,
+            out string errorMessage)
+        {
+            object property = null;
+            if (!TryGetDynamicObjectProperty(target, contractResolver, segment, out property))
+            {
+                errorMessage = Resources.FormatTargetLocationAtPathSegmentNotFound(segment);
+                return false;
+            }
+
+            object convertedValue = null;
+            if (!TryConvertValue(value, property.GetType(), out convertedValue))
+            {
+                errorMessage = Resources.FormatInvalidValueForProperty(value);
+                return false;
+            }
+
+            if (!TrySetDynamicObjectProperty(target, contractResolver, segment, convertedValue))
+            {
+                errorMessage = Resources.FormatTargetLocationAtPathSegmentNotFound(segment);
+                return false;
+            }
+
+            errorMessage = null;
+            return true;
+        }
+
+        public bool TryTraverse(
+            object target,
+            string segment,
+            IContractResolver contractResolver,
+            out object nextTarget,
+            out string errorMessage)
+        {
+            var dynamicObject = target as IDynamicMetaObjectProvider;
+            if (dynamicObject == null)
+            {
+                nextTarget = null;
+                errorMessage = null;
+                return false;
+            }
```

Comment:
Under what conditions can this happen? Should we be checking this in all of these methods and returning a good error message?